### DISCUSSION TITLE
Can't connect to linux localhost from Windows browser

### DISCUSSION THREAD

**Discussion by programmylife** [microsoft/WSL#2471] [General]
I am trying to run two localhost applications (not at the same time) in bash on Ubuntu on Windows: flask and jekyll. I'm starting the services from the bash on Windows terminal using 'jekyll serve' and 'python -m flask run'. In both cases I copy and paste the localhost address printed to the terminal (127.0.0.1:5000, and 127.0.0.1:4000, respectively) to a browser on Windows 10, and get 'This site can't be reached'.

Edit: I have installed flask on windows and am able to run and access a local server from the Windows command line as expected (but still not bash).

I have tried upgrading to the newer version of Ubuntu (uninstalled and reinstalled to go from v14 to v16) as well as creating a firewall rule for the bash.exe file on Windows Defender.

Microsoft Windows build number [Version 10.0.15063]

**Comment by gravcat** [comment]
If you telnet to that port from within the Linux SubSystem does it connect, or fail?

`telnet localhost 5000`
`telnet localhost 4000`

**Comment by programmylife** [comment]
Fails. I tried it in a separate window and the same window (both bash).

**Comment by gravcat** [comment]
If it doesn't work in bash it makes me think you have a firewall restriction of some sort, can you try just blanket rule shutting the firewall off to see? Also do `sudo ufw status` to make sure Ubuntu isn't running a firewall.

**Comment by programmylife** [comment]
Thanks, Nick. I'm out of town for the weekend, but I'll try those on Monday
when I get back and report here.

On Sep 8, 2017 12:11 PM, "Nick Thieling" <notifications@github.com> wrote:

If it doesn't work in bash it makes me think you have a firewall
restriction of some sort, can you try just blanket rule shutting the
firewall off to see? Also do sudo ufw status to make sure Ubuntu isn't
running a firewall.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/Microsoft/BashOnWindows/issues/2471#issuecomment-328147031>,
or mute the thread

**Comment by programmylife** [comment]
Disabling the Windows Defender Firewalls didn't work. Trying to run 'sudo ufw status ' gave me an error:

ERROR: problem running iptables: iptables v1.6.0: can't initialize iptables table `filter': Table does not exist (do you need to insmod?)
Perhaps iptables or your kernel needs to be upgraded.

**Comment by gravcat** [comment]
Silly question, but if you do `ps -ef | grep python` do you see the apps running and not in a failed state? Can you check your logging to confirm it's working properly?

Perhaps the problem is simpler than we think...

**Comment by programmylife** [comment]
'acmshar     12     2 10 07:44 tty1     00:00:00 python -m flask run'

So it appears to be running.

**Comment by hybridtechie** [comment]
Hi 

I am trying to get Memcached running on WSL and try to access it via the port in Windows browser.
It says `Could not open port.`
I am getting the same error when trying to `sudo ufw status`
I am also facing the same issue and python seems to be running.

**Comment by matteocontrini** [comment]
Same here with a Node.js HTTP server.

winver => 15063.608

I think this is a duplicate of #1498

**Comment by Wvvi** [comment]
This is a problem that bothers me, too，Would you like to have a solution?

**Comment by karenyyng** [comment]
WSL runs via Docker like technology and has the same restrictions.
You have to point your web app inside the WSL to use the IP ‘0.0.0.0’ and then you can see the app in a browser at 127.0.0.1

**Reply by luisdev** [reply]
"You have to point your web app inside the WSL to use the IP ‘0.0.0.0’ " - How do you do that?

**Comment by kc2gug** [comment]
@karenyyng
Best answer thus far, though it did not resolve my issue. 
     FLASK_APP=basic.py flask run -h 0.0.0.0 -p 12345

Simple program to test with

```
import flask

app = flask.Flask(__name__)


@app.route('/')
def hello_world():
    return 'Hello, World!'
```

Any thoughts?

**Comment by muhammad-amhan** [comment]
I have exactly the same problem too, I'm on Windows 7. 

So my Flask app runs on `0.0.0.0:5000`. Using **Bash on Windows**, I dockerized the app and started the service from a docker-compose file.  I cannot access it through Windows Chrome browser, same error message `This site can't be reached`.

If I run the Flask app from my IDE (IntelliJ) on Windows (outside of Bash), then I can simply access the API through Chrome on `127.0.0.1:5000/test` for example. 

I also disabled the firewall completely for testing purposes and it didn't solve it. I'm not sure how to tell Bash that the localhost the service is starting on, is the machine's localhost. Because I started the app through IntelliJ at the same time, using the same host and port, and it successfully started although the service in my Bash is running on the same host:port too!! How can that be possible? I would think they all refer to the same machine localhost.

**Comment by muhammad-amhan** [comment]
Update: 

I figured out what's happening and it all makes sense now. It's my first time using `Docker ToolBox` not knowing that it is different to `Docker for Windows`.  So whenever I'm running services using docker in Bash, it's not using the Windows localhost despite `docker-compose logs <service name>` shows that service is running on `0.0.0.0:5000` it's in fact running on `<docker machine ip address>:5000` in my case because I set the port to **5000**.

You can grab the docker machine ip address simply by running `docker-machine ip` and use it as follow:
`<ip>:<port>` in your windows browser to access the service running by docker-compose.

**Reply by RJStevenson** [reply]
thanks, @muhammad-amhan  this helped me a lot. 

it's so weird if I restart my computer and immediately launch the node server it works on port 3000.
but then when I closed it and restarted the server it didn't.

your fix. really helped.
thanks again

**Comment by clshortfuse** [comment]
This started happening to me with WSL 2 on Windows 10-18917. Never had issues before.

Edit: Downgraded to WSL 1 and it works fine. WSL 2 doesn't.

Edit2: Didn't realize support for it was dropped in WSL2. That's a deal-breaker for me.

https://docs.microsoft.com/en-us/windows/wsl/wsl2-ux-changes#accessing-network-applications

**Comment by benhillis** [comment]
@clshortfuse - We are working on enhancing our networking story to allow connecting to services running in the guest via localhost in a future update.

**Comment by Gorthog** [comment]
Seems this support is live in build 18945, but still not working for me.
Serving Angular app inside WSL 2, works fine (wget localhost:4200 downloads index.html). While trying wget from Windows returns unable to connect. I'm running build 18950.1000. Any help / workaround would be vastly appreciated.

**Comment by Gorthog** [comment]
seems my real issue is https://github.com/microsoft/WSL/issues/4353

**Comment by j-collier** [comment]
You can fix this issue by updating your Windows hosts file.

In a shell in WSL 2 Ubuntu run `$ip addr` to find the VM's IP address. Then edit the Window's host file at `C:\Windows\System32\drivers\etc\hosts` and add the IP address and a name at the bottom of the file. For example: `111.22.333.444 wsl` which will allow you to access processes serving to localhost in WSL 2 by navigating to http://wsl:{your_port_number}

**Reply by danuvian** [reply]
This worked for me. Thank you!

**Reply by givihuda** [reply]
:/mnt/c/Users/me$ $ip addr

Command 'addr' not found, did you mean:

  command 'maddr' from deb mblaze (0.6-1)
  command 'addr6' from deb ipv6toolkit (2.0-1)
  command 'addcr' from deb courier-mta (1.0.6-1build2)
  command 'addcr' from deb ucspi-tcp (1:0.88-6)
  command 'addcr' from deb ucspi-tcp-ipv6 (1:0.88-6)

Try: sudo apt install <deb name>

Which one should I install? I'm using WSL2 Ubuntu 20.

**Reply by j-collier** [reply]
ah that `$` is supposed to symbolize your prompt, not something to include in the command. Just run:
```
ip addr
```

**Reply by givihuda** [reply]
Understood... typically the `$` is separated from the command text by a space though even when referenced, as `$` being attached to text literally represents a var in Linux. I'd recommend editing your original post to separate the space from the dollar symbol. Thanks for coming to my Ted Talk.

**Reply by pravinweb** [reply]
Example - VMs IP address after running ip addr

The IP address for eth0 is **10.1.0.143.**

2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc mq state UP group default qlen 1000
link/ether 02:3f:4d:eb:52:e7 brd ff:ff:ff:ff:ff:ff
inet **10.1.0.143**/24 brd 10.1.0.255 scope global dynamic noprefixroute eth0
valid_lft 3198sec preferred_lft 3198sec


Thanks to - https://www.howtouselinux.com/post/check-ip-address-in-linux

**Comment by mickdewald** [comment]
`wsl --shutdown` helped me in Win 2004

**Reply by chicagobuss** [reply]
thank you, this worked for me, too

**Reply by edwinmartinezplus** [reply]
I cannot tell you how many hours i wasted until i found this one line that solved it all. Thank you.

**Reply by dqgiang** [reply]
Thank you. This works for me.

I was using `localhost:port` just fine in Windows version 2004, with both WSL1 and WSL2.
After updated to version 20H2 Build 19042.685 today with a few reboots I got into this issue. Restarting WSL helped.
Docker was still running normally though. 😕

**Reply by samsimsom** [reply]
thank you.

**Reply by givihuda** [reply]
Windows should be running that when you shut down, not brute forcing an exit and causing the error. 

I already have to click and type enough crap to get things to shut down properly in all of the other development software I have to use.

**Reply by antonstjernquist** [reply]
This is still the only working solution for me too :)

**Reply by dongocanh272** [reply]
Thank you! This worked for me. Just run `wsl --shutdown` and start everything again :D

**Reply by samazevedo** [reply]
Thank you, it worked

**Reply by Marcel-MSC** [reply]
> `wsl --shutdown` helped me in Win 2004

It worked for me too. I'm using Windows 10... 
I don't know if the mongodb frozen inside wls2 or it's a bug.
2 hours of my day trying to understand this shit.

**Reply by DumiduPramith** [reply]
> `wsl --shutdown` helped me in Win 2004

thanks lot

**Reply by german1311** [reply]
worked, I'm using w10 and wsl2,

Run:
`wsl`
to start again in the same terminal

**Reply by Weeelliam** [reply]
yeah, this!

**Reply by nauman-CSsol** [reply]
Works. Thanks!
wsl --shutdown
Then restarting my pc helped me as well. Now I can run my all programs: localhost:4200, localhost:8080 and localhost:8000.
I am using Windows 21H1 and wsl2.

**Reply by srph** [reply]
Ugh, thank you!

**Reply by bcarlog** [reply]
Yeah it works!

**Reply by vpeopleonatank** [reply]
Yeah it work! Thank you so much!

**Reply by MNaumanBootter** [reply]
It works! Thanks.

**Reply by ppostnov** [reply]
Thanks a lot ! Works for me too.

**Reply by 0fprod** [reply]
Yup, you're the hero of the day :) 
Thanks

**Reply by moranlemusj** [reply]
You sir, get a gold star ⭐ 

Thank you so much

**Comment by jssoriao** [comment]
I also have experienced this where initially I can run an api server in wsl2 and call its endpoints using postman (localhost) running in the host machine. But after some time, it just stops working and I get connection error. But after running `wsl --shutdown`, and restarting wsl2, it works again. A fix would be much better than this workaround

**Reply by Vivek-Py** [reply]
After wasting hours, all it came to `wsl --shutdown`. I want to laugh but I can't.

**Reply by DispatchCommit** [reply]
Just updated to WSL2, had no issues, rebooted my PC and found I couldn't connect to my node.js dev server running in WSL2...

`wsl --shutdown` and restarting solved it for me as well. This was just.. frustrating..

**Reply by ilijapet** [reply]
Works for me as well! 
wsl --shutdown
Then restarting wsl and everything works just fine 
I can then start localhost in wsl with lets say: python3 -m http.server 8000 
and then from browser: localhost:8000  
Everything there!
Thanks @micktg!!! I waist 4 hours but you save to us much more :))

**Reply by skovy** [reply]
I had to run `wsl.exe --shutdown` but this did the trick for me!

**Comment by Amondale** [comment]
@jsoriao, I appreciate hearing about your workaround, it works for me; WSL2 has one annoyance that may have roots in Docker WSL2 integration: mapping kubernetes to 127.0.0.1 in the *Windows* hosts file, then that same mapping is transferred to /etc/hosts in BASH. I believe you can turn off the sync between BASH and Windows HOSTS static mappings with a change to /etc/wsl.conf noted in the hosts file headers. Perhaps this is the way to prevent Docker from closing connections on 127.0.0.1? I don't use Docker extensively so not sure. I would check out "Network", "Proxy" and "WSL Integration" settings under Resources in Docker Dashboard, they seem to impact when Windows stops being able to connect to BASH on 127.0.0.1.

**Comment by Amondale** [comment]
> You can fix this issue by updating your Windows hosts file.
> 
> In a shell in WSL 2 Ubuntu run `$ip addr` to find the VM's IP address. Then edit the Window's host file at `C:\Windows\System32\drivers\etc\hosts` and add the IP address and a name at the bottom of the file. For example: `111.22.333.444 wsl` which will allow you to access processes serving to localhost in WSL 2 by navigating to [http://wsl:{your_port_number}](http://wsl:%7Byour_port_number%7D)

problem for me is that the IPv4 address seems to be "dynamic" between restarts of WSL2, so entering it in a static HOSTS file is a bit like the dog chasing its tail. I'd rather track down what causes the CONNECTION_REFUSED in my windows browser. I am pretty sure it is sporadic on one particular 2004 instance, on a different one, I have never had this problem, although pretty much the same basic configuration: Docker, WSL2, nginx within WSL2 as a proxy, attaching to :80 on the BASH host (nginx proxying a Django Web App).

**Comment by Will-t-future** [comment]
hi I have a problem that when I want to use vscode to connect WSL2,there is something wrong `/root/.vscode-server/bin/93c2f0fbf16c5a4b10e4d5f89737d9c2c25488a3/server.sh: line 12: /root/.vscode-server/bin/93c2f0fbf16c5a4b10e4d5f89737d9c2c25488a3/node: not found`
what can I do to fix it

**Comment by ghost** [comment]
Do you have a .wslconfig file?

If no, create one with `localhostForwarding=true`:

An example of .wslconfig file:
```
[wsl2]
memory=4GB
swap=0
localhostForwarding=true
```

if yes,

make sure that:
`localhostForwarding=true`

**Reply by kimlym** [reply]
This used to work for me, but stops on Windows Build 20H2 19042.608. I have to use ip addr to access now. However, my docker apps are not affected

**Reply by kaisellgren** [reply]
The documentation said this is the default (`true`), but as soon as I set it to `true` manually in the config and restarted wsl (`wsl --shutdown && wsl`) then it started to work.

**Reply by Grawl** [reply]
Thank you!

Here's more info about: https://docs.microsoft.com/en-us/windows/wsl/wsl-config#configure-global-options-with-wslconfig

I just added it to my article on using WSL + Docker https://dev.to/grawl/working-with-docker-and-other-unix-things-on-windows-like-on-macos-or-linux-4o6p

**Comment by robertmielewczyk** [comment]
I solved my issue by using ifconfig command and getting IP from there

**What I was doing:**
Running Haddop from WSL: Ubuntu

**Steps I've taken**
1. started application with -host 0.0.0.0 -port 8080
2. looked for ip with ifconfig command
3. open url with {ip_from_ifconfig}:{port}

**Comment by SeanChao** [comment]
After listening a port in WSL, go to windows and run `netstat -ano | findstr PORT`.

For me, my program actually listened at `::1`, while I was expecting to access it from `127.0.0.1`.

**Comment by j-collier** [comment]
For everyone still finding this thread: Windows Insider Preview build 21292.1010 fixes this issue and you no longer need to proxy or set a custom host.

**Reply by luisdev** [reply]
Thanks. It's now July 2021, has this fix been released yet? If so, what's the Windows build number for the release?

**Comment by basaran** [comment]
I recently started having this issue. During development, sometimes the `localhost` connections from `windows` to `wsl` would just stop working. Standard `wsl -t ubuntu` would not restore `localhost`. I would have to restart the `lxssmanager` service to get localhost back.

While the `windows` to `wsl` `localhost` is gone:

1. the ip from ifconfig connects fine from windows. `http://172.17.74.198....`
2. to ensure it's not my dev server, I opened a port with `nc -l 5000` and it would fail. `telnet localhost 5000` , `telnet 127.0.0.1  5000` 
3. During all this time, WSL has no issue connecting to internet.
4. `netstat -ano` only returns the port for the `ifconfig` ip.
5. restarting the dev server inside wsl has no effect to bindings.

If there is anything else you need to me to check next time this happens, please let me know.

**Reply by icarus42** [reply]
Problem appeared for me after a Windows restart out of nowhere (I am not aware of any changes that could have caused the connection from Windows host to services inside WSL not to work). Neither `wsl -t <distro>` nor rebooting Windows would fix it, but **restarting the `lxssmanager`** service did it for me. Thank you 👍

**Reply by lexuzieel** [reply]
For some reason restarting `lxssmanager` service didn't work for me. This issue appears randomly on Windows boot and only rebooting fixes it for me currently.

**Reply by Marcel-MSC** [reply]
haver you guys try use cmd to shutdown wsl instances that can be on?
open cmd and type wsl --shutdown

**Comment by tom-cc** [comment]
Hi!
This fixes the issue of accessing web servers running on WSL2 from Windows: https://docs.microsoft.com/en-us/windows/wsl/compare-versions#accessing-a-wsl-2-distribution-from-your-local-area-network-lan

**Comment by sudomaxime** [comment]
I fixed my issue by adding this to my windows host machine host-file:

You can find it at C:\Windows\System32\drivers\etc\hosts

To edit it, press the windows button on your keyboard, type "notepad" and right click to "open as administrator". 

At the bottom of the file add this

```
::1 localhost
::1 whatever-subdomain-you-may-want-to-use-here.localhost
```

**Comment by timeneverdie** [comment]
I guess you could try to disable Chrome Sandbox UWP Loopback by `Fiddler`.

Open `Fiddler`, click `WinConfig`, then check `Chrome Sandbox`. Save changes.

Try to disable WSL relative items too.

**Comment by HonzaKopecky** [comment]
Recently, I started to experience an issue where the `::1 whatever-domain` rule wouldn't be enough for Chromium browsers. Adding a `127.0.0.1 whatever-domain` rule helped so my current setup for each domain is as follows:

```
::1 whatever-domain1
127.0.0.1 whatever-domain1

::1 whatever-domain2
127.0.0.1 whatever-domain2

...

```
